sql/pgwire: TestPGTest failed

sql/pgwire.TestPGTest [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4443087&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4443087&tab=artifacts#/) on master @ [1f4dde2c5fdccf831e1fc4abc210df507350f8fa](https://github.com/cockroachdb/cockroach/commits/1f4dde2c5fdccf831e1fc4abc210df507350f8fa):


```
RECV *pgproto3.ParameterStatus: &{Name:integer_datetimes Value:on}
RECV *pgproto3.ParameterStatus: &{Name:standard_conforming_strings Value:on}
RECV *pgproto3.ParameterStatus: &{Name:crdb_version Value:CockroachDB OSS v22.1.0-alpha.1-1607-g1f4dde2c5f (x86_64-pc-linux-gnu, built , go1.17.6)}
RECV *pgproto3.ParameterStatus: &{Name:session_authorization Value:root}
RECV *pgproto3.BackendKeyData: &{ProcessID:2080058 SecretKey:3649054758}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:SELECT pg_sleep(0.1)}
RECV *pgproto3.RowDescription: &{Fields:[{Name:[112 103 95 115 108 101 101 112] TableOID:0 TableAttributeNumber:0 DataTypeOID:16 DataTypeSize:1 TypeModifier:-1 Format:0}]}
RECV *pgproto3.DataRow: &{Values:[[116]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:CREATE TABLE tab(a INT8)}
RECV *pgproto3.CommandComplete: &{CommandTag:[67 82 69 65 84 69 32 84 65 66 76 69]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Parse: &{Name:s0 Query:SELECT * FROM tab AS OF SYSTEM TIME '-0.05s' ParameterOIDs:[]}
SEND *pgproto3.Sync: &{}
RECV *pgproto3.ErrorResponse: &{Severity:ERROR SeverityUnlocalized: Code:42P01 Message:relation "tab" does not exist Detail: Hint: Position:0 InternalPosition:0 InternalQuery: Where: SchemaName: TableName: ColumnName: DataTypeName: ConstraintName: File:errors.go Line:161 Routine:NewUndefinedRelationError UnknownFields:map[]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:SELECT pg_sleep(0.1)}
RECV *pgproto3.RowDescription: &{Fields:[{Name:[112 103 95 115 108 101 101 112] TableOID:0 TableAttributeNumber:0 DataTypeOID:16 DataTypeSize:1 TypeModifier:-1 Format:0}]}
RECV *pgproto3.DataRow: &{Values:[[116]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:INSERT INTO tab VALUES(1)}
RECV *pgproto3.CommandComplete: &{CommandTag:[73 78 83 69 82 84 32 48 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Parse: &{Name:historical_stmt Query:SELECT * FROM tab AS OF SYSTEM TIME '-0.05s' ParameterOIDs:[]}
SEND *pgproto3.Bind: &{DestinationPortal:p1 PreparedStatement:historical_stmt ParameterFormatCodes:[] Parameters:[] ResultFormatCodes:[]}
SEND *pgproto3.Execute: &{Portal:p1 MaxRows:0}
SEND *pgproto3.Sync: &{}
RECV *pgproto3.ParseComplete: &{}
RECV *pgproto3.BindComplete: &{}
RECV *pgproto3.DataRow: &{Values:[[49]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
    datadriven.go:75: 
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:72: ReadyForQuery
        expected:
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"CommandComplete","CommandTag":"SELECT 0"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
        
        found:
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"DataRow","Values":[{"text":"1"}]}
        {"Type":"CommandComplete","CommandTag":"SELECT 1"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
    --- FAIL: TestPGTest/as_of_system_time (1.97s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-experience
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPGTest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/pgwire.TestPGTest [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445217&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445217&tab=artifacts#/) on master @ [a7449866f3229872d92e1de973e209838f07de35](https://github.com/cockroachdb/cockroach/commits/a7449866f3229872d92e1de973e209838f07de35):


```
RECV *pgproto3.ParameterStatus: &{Name:integer_datetimes Value:on}
RECV *pgproto3.ParameterStatus: &{Name:standard_conforming_strings Value:on}
RECV *pgproto3.ParameterStatus: &{Name:crdb_version Value:CockroachDB OSS v22.1.0-alpha.1-1614-ga7449866f3 (x86_64-pc-linux-gnu, built , go1.17.6)}
RECV *pgproto3.ParameterStatus: &{Name:session_authorization Value:root}
RECV *pgproto3.BackendKeyData: &{ProcessID:1478239 SecretKey:3241886040}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:SELECT pg_sleep(0.1)}
RECV *pgproto3.RowDescription: &{Fields:[{Name:[112 103 95 115 108 101 101 112] TableOID:0 TableAttributeNumber:0 DataTypeOID:16 DataTypeSize:1 TypeModifier:-1 Format:0}]}
RECV *pgproto3.DataRow: &{Values:[[116]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:CREATE TABLE tab(a INT8)}
RECV *pgproto3.CommandComplete: &{CommandTag:[67 82 69 65 84 69 32 84 65 66 76 69]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Parse: &{Name:s0 Query:SELECT * FROM tab AS OF SYSTEM TIME '-0.05s' ParameterOIDs:[]}
SEND *pgproto3.Sync: &{}
RECV *pgproto3.ErrorResponse: &{Severity:ERROR SeverityUnlocalized: Code:42P01 Message:relation "tab" does not exist Detail: Hint: Position:0 InternalPosition:0 InternalQuery: Where: SchemaName: TableName: ColumnName: DataTypeName: ConstraintName: File:errors.go Line:161 Routine:NewUndefinedRelationError UnknownFields:map[]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:SELECT pg_sleep(0.1)}
RECV *pgproto3.RowDescription: &{Fields:[{Name:[112 103 95 115 108 101 101 112] TableOID:0 TableAttributeNumber:0 DataTypeOID:16 DataTypeSize:1 TypeModifier:-1 Format:0}]}
RECV *pgproto3.DataRow: &{Values:[[116]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:INSERT INTO tab VALUES(1)}
RECV *pgproto3.CommandComplete: &{CommandTag:[73 78 83 69 82 84 32 48 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Parse: &{Name:historical_stmt Query:SELECT * FROM tab AS OF SYSTEM TIME '-0.05s' ParameterOIDs:[]}
SEND *pgproto3.Bind: &{DestinationPortal:p1 PreparedStatement:historical_stmt ParameterFormatCodes:[] Parameters:[] ResultFormatCodes:[]}
SEND *pgproto3.Execute: &{Portal:p1 MaxRows:0}
SEND *pgproto3.Sync: &{}
RECV *pgproto3.ParseComplete: &{}
RECV *pgproto3.BindComplete: &{}
RECV *pgproto3.DataRow: &{Values:[[49]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
    datadriven.go:75: 
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:72: ReadyForQuery
        expected:
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"CommandComplete","CommandTag":"SELECT 0"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
        
        found:
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"DataRow","Values":[{"text":"1"}]}
        {"Type":"CommandComplete","CommandTag":"SELECT 1"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
    --- FAIL: TestPGTest/as_of_system_time (2.12s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPGTest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/pgwire.TestPGTest [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4446848&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4446848&tab=artifacts#/) on master @ [a04ecc88a678464fab1982cde3d1b2f120b1b126](https://github.com/cockroachdb/cockroach/commits/a04ecc88a678464fab1982cde3d1b2f120b1b126):


```
SEND *pgproto3.Execute: &{Portal:p1 MaxRows:0}
SEND *pgproto3.Sync: &{}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1299/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:65:
        send [1 args]
        Parse {"Name": "historical_stmt", "Query": "SELECT * FROM tab AS OF SYSTEM TIME '-0.05s'"}
        Bind {"DestinationPortal": "p1", "PreparedStatement": "historical_stmt"}
        Execute {"Portal": "p1"}
        Sync
        ----
RECV *pgproto3.ParseComplete: &{}
RECV *pgproto3.BindComplete: &{}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 48]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1299/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:72:
        until [1 args]
        ReadyForQuery
        ----
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"CommandComplete","CommandTag":"SELECT 0"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
SEND *pgproto3.Bind: &{DestinationPortal:p2 PreparedStatement:historical_stmt ParameterFormatCodes:[] Parameters:[] ResultFormatCodes:[]}
SEND *pgproto3.Execute: &{Portal:p2 MaxRows:0}
SEND *pgproto3.Sync: &{}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1299/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:82:
        send [1 args]
        Bind {"DestinationPortal": "p2", "PreparedStatement": "historical_stmt"}
        Execute {"Portal": "p2"}
        Sync
        ----
RECV *pgproto3.BindComplete: &{}
RECV *pgproto3.DataRow: &{Values:[[49]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1299/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:88: ReadyForQuery
        expected:
        {"Type":"BindComplete"}
        {"Type":"CommandComplete","CommandTag":"SELECT 0"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
        
        found:
        {"Type":"BindComplete"}
        {"Type":"DataRow","Values":[{"text":"1"}]}
        {"Type":"CommandComplete","CommandTag":"SELECT 1"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
    --- FAIL: TestPGTest/as_of_system_time (0.97s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPGTest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/pgwire.TestPGTest [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4446849&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4446849&tab=artifacts#/) on master @ [a04ecc88a678464fab1982cde3d1b2f120b1b126](https://github.com/cockroachdb/cockroach/commits/a04ecc88a678464fab1982cde3d1b2f120b1b126):


```
SEND *pgproto3.Execute: &{Portal:p1 MaxRows:0}
SEND *pgproto3.Sync: &{}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1117/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:65:
        send [1 args]
        Parse {"Name": "historical_stmt", "Query": "SELECT * FROM tab AS OF SYSTEM TIME '-0.05s'"}
        Bind {"DestinationPortal": "p1", "PreparedStatement": "historical_stmt"}
        Execute {"Portal": "p1"}
        Sync
        ----
RECV *pgproto3.ParseComplete: &{}
RECV *pgproto3.BindComplete: &{}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 48]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1117/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:72:
        until [1 args]
        ReadyForQuery
        ----
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"CommandComplete","CommandTag":"SELECT 0"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
SEND *pgproto3.Bind: &{DestinationPortal:p2 PreparedStatement:historical_stmt ParameterFormatCodes:[] Parameters:[] ResultFormatCodes:[]}
SEND *pgproto3.Execute: &{Portal:p2 MaxRows:0}
SEND *pgproto3.Sync: &{}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1117/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:82:
        send [1 args]
        Bind {"DestinationPortal": "p2", "PreparedStatement": "historical_stmt"}
        Execute {"Portal": "p2"}
        Sync
        ----
RECV *pgproto3.BindComplete: &{}
RECV *pgproto3.DataRow: &{Values:[[49]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
    datadriven.go:75: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1117/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:88: ReadyForQuery
        expected:
        {"Type":"BindComplete"}
        {"Type":"CommandComplete","CommandTag":"SELECT 0"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
        
        found:
        {"Type":"BindComplete"}
        {"Type":"DataRow","Values":[{"text":"1"}]}
        {"Type":"CommandComplete","CommandTag":"SELECT 1"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
    --- FAIL: TestPGTest/as_of_system_time (1.81s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPGTest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/pgwire.TestPGTest [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4448492&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4448492&tab=artifacts#/) on master @ [653e35c0b1dabd8c2d4827f39dfc20da6ecf10d0](https://github.com/cockroachdb/cockroach/commits/653e35c0b1dabd8c2d4827f39dfc20da6ecf10d0):


```
RECV *pgproto3.ParameterStatus: &{Name:integer_datetimes Value:on}
RECV *pgproto3.ParameterStatus: &{Name:standard_conforming_strings Value:on}
RECV *pgproto3.ParameterStatus: &{Name:crdb_version Value:CockroachDB OSS v22.1.0-alpha.1-1622-g653e35c0b1 (x86_64-pc-linux-gnu, built , go1.17.6)}
RECV *pgproto3.ParameterStatus: &{Name:session_authorization Value:root}
RECV *pgproto3.BackendKeyData: &{ProcessID:1523099 SecretKey:1681013296}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:SELECT pg_sleep(0.1)}
RECV *pgproto3.RowDescription: &{Fields:[{Name:[112 103 95 115 108 101 101 112] TableOID:0 TableAttributeNumber:0 DataTypeOID:16 DataTypeSize:1 TypeModifier:-1 Format:0}]}
RECV *pgproto3.DataRow: &{Values:[[116]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:CREATE TABLE tab(a INT8)}
RECV *pgproto3.CommandComplete: &{CommandTag:[67 82 69 65 84 69 32 84 65 66 76 69]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Parse: &{Name:s0 Query:SELECT * FROM tab AS OF SYSTEM TIME '-0.05s' ParameterOIDs:[]}
SEND *pgproto3.Sync: &{}
RECV *pgproto3.ErrorResponse: &{Severity:ERROR SeverityUnlocalized: Code:42P01 Message:relation "tab" does not exist Detail: Hint: Position:0 InternalPosition:0 InternalQuery: Where: SchemaName: TableName: ColumnName: DataTypeName: ConstraintName: File:errors.go Line:161 Routine:NewUndefinedRelationError UnknownFields:map[]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:SELECT pg_sleep(0.1)}
RECV *pgproto3.RowDescription: &{Fields:[{Name:[112 103 95 115 108 101 101 112] TableOID:0 TableAttributeNumber:0 DataTypeOID:16 DataTypeSize:1 TypeModifier:-1 Format:0}]}
RECV *pgproto3.DataRow: &{Values:[[116]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Query: &{String:INSERT INTO tab VALUES(1)}
RECV *pgproto3.CommandComplete: &{CommandTag:[73 78 83 69 82 84 32 48 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
SEND *pgproto3.Parse: &{Name:historical_stmt Query:SELECT * FROM tab AS OF SYSTEM TIME '-0.05s' ParameterOIDs:[]}
SEND *pgproto3.Bind: &{DestinationPortal:p1 PreparedStatement:historical_stmt ParameterFormatCodes:[] Parameters:[] ResultFormatCodes:[]}
SEND *pgproto3.Execute: &{Portal:p1 MaxRows:0}
SEND *pgproto3.Sync: &{}
RECV *pgproto3.ParseComplete: &{}
RECV *pgproto3.BindComplete: &{}
RECV *pgproto3.DataRow: &{Values:[[49]]}
RECV *pgproto3.CommandComplete: &{CommandTag:[83 69 76 69 67 84 32 49]}
RECV *pgproto3.ReadyForQuery: &{TxStatus:73}
    datadriven.go:75: 
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/testdata/pgtest/as_of_system_time:72: ReadyForQuery
        expected:
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"CommandComplete","CommandTag":"SELECT 0"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
        
        found:
        {"Type":"ParseComplete"}
        {"Type":"BindComplete"}
        {"Type":"DataRow","Values":[{"text":"1"}]}
        {"Type":"CommandComplete","CommandTag":"SELECT 1"}
        {"Type":"ReadyForQuery","TxStatus":"I"}
    --- FAIL: TestPGTest/as_of_system_time (0.87s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPGTest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/pgwire/testdata/pgtest/as_of_system_time
+++ b/pkg/sql/pgwire/testdata/pgtest/as_of_system_time
@@ -1,23 +1,18 @@
 # This test relies on a CoockroachDB-specific feature, so everything
 # is marked as crdb_only.
 
-send crdb_only
-Query {"String": "SELECT pg_sleep(0.1)"}
+only crdb
 ----
 
-until crdb_only
-ReadyForQuery
+let $before_table_timestamp
+Query {"String": "SELECT cluster_logical_timestamp()"}
 ----
-{"Type":"RowDescription","Fields":[{"Name":"pg_sleep","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":16,"DataTypeSize":1,"TypeModifier":-1,"Format":0}]}
-{"Type":"DataRow","Values":[{"text":"t"}]}
-{"Type":"CommandComplete","CommandTag":"SELECT 1"}
-{"Type":"ReadyForQuery","TxStatus":"I"}
 
-send crdb_only
+send
 Query {"String": "CREATE TABLE tab(a INT8)"}
 ----
 
-until crdb_only
+until
 ReadyForQuery
 ----
 {"Type":"CommandComplete","CommandTag":"CREATE TABLE"}
@@ -25,37 +20,96 @@ ReadyForQuery
 
 # Make sure AOST is handled during Parse. Preparing the statement should
 # fail since the table did not exist in the past.
-send crdb_only
-Parse {"Name": "s0", "Query": "SELECT * FROM tab AS OF SYSTEM TIME '-0.05s'"}
+send
+Parse {"Name": "s0", "Query": "SELECT * FROM tab AS OF SYSTEM TIME '$before_table_timestamp'"}
 Sync
 ----
 
-until crdb_only
+until keepErrMessage
 ErrorResponse
 ReadyForQuery
 ----
-{"Type":"ErrorResponse","Code":"42P01"}
+{"Type":"ErrorResponse","Code":"42P01","Message":"relation \"tab\" does not exist"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
 
-send crdb_only
-Query {"String": "SELECT pg_sleep(0.1)"}
+let $before_data_timestamp
+Query {"String": "SELECT cluster_logical_timestamp()"}
 ----
 
-until crdb_only
-ReadyForQuery
-----
-{"Type":"RowDescription","Fields":[{"Name":"pg_sleep","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":16,"DataTypeSize":1,"TypeModifier":-1,"Format":0}]}
-{"Type":"DataRow","Values":[{"text":"t"}]}
-{"Type":"CommandComplete","CommandTag":"SELECT 1"}
-{"Type":"ReadyForQuery","TxStatus":"I"}
-
-
-send crdb_only
+send
 Query {"String": "INSERT INTO tab VALUES(1)"}
 ----
 
-until crdb_only
+until
 ReadyForQuery
 ----
 {"Type":"CommandComplete","CommandTag":"INSERT 0 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Make sure AOST is handled consistently during Parse/Bind/Execute. This should
+# succeed, but should not be able to read the data that was added to the table.
+send
+Parse {"Name": "historical_stmt", "Query": "SELECT * FROM tab AS OF SYSTEM TIME '$before_data_timestamp'"}
+Bind {"DestinationPortal": "p1", "PreparedStatement": "historical_stmt"}
+Execute {"Portal": "p1"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"CommandComplete","CommandTag":"SELECT 0"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Make sure AOST is handled consistently during Bind/Execute. This also should
+# not be able to see the data in the table, since this is a historical read.
+send
+Bind {"DestinationPortal": "p2", "PreparedStatement": "historical_stmt"}
+Execute {"Portal": "p2"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"BindComplete"}
+{"Type":"CommandComplete","CommandTag":"SELECT 0"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+let $after_data_timestamp
+Query {"String": "SELECT cluster_logical_timestamp()"}
+----
+
+# Preparing another statement with the later timestamp should succeed.
+send
+Parse {"Name": "historical_stmt_working", "Query": "SELECT * FROM tab AS OF SYSTEM TIME '$after_data_timestamp'"}
+Bind {"DestinationPortal": "p3", "PreparedStatement": "historical_stmt_working"}
+Execute {"Portal": "p3"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"1"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# And reusing that statement should work as well.
+send
+Bind {"DestinationPortal": "p4", "PreparedStatement": "historical_stmt_working"}
+Execute {"Portal": "p4"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"1"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
 {"Type":"ReadyForQuery","TxStatus":"I"}

--- a/pkg/sql/pgwire/testdata/pgtest/bind_and_resolve
+++ b/pkg/sql/pgwire/testdata/pgtest/bind_and_resolve
@@ -1,10 +1,6 @@
-# Test binding calls which reference names via either types or oid casts work
-# after a bizarre turn of events whereby we used to set the planner txn to nil.
-# The long and the short of it is that we totally abuse the planner when
-# performing BIND and the fact that it continues to carry a transaction that
-# does not correspond to the connExecutor's transaction is, generally, a big
-# problem. Before the changes made in the commit adding this test, this test
-# would result in a nil pointer panic.
+# Test binding calls which reference names via either types or oid casts work.
+# Before the changes made in the commit adding this test, this test would result
+# in a nil pointer panic, since the txn was not initialized correctly.
 
 # TODO(richardjcai): Support let command similar to logic tests to
 # programmatically get the table id.

--- a/pkg/testutils/pgtest/BUILD.bazel
+++ b/pkg/testutils/pgtest/BUILD.bazel
@@ -13,5 +13,6 @@ go_library(
         "@com_github_cockroachdb_datadriven//:datadriven",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_jackc_pgproto3_v2//:pgproto3",
+        "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/testutils/pgtest/datadriven.go
+++ b/pkg/testutils/pgtest/datadriven.go
@@ -20,6 +20,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/datadriven"
 	"github.com/jackc/pgproto3/v2"
+	"github.com/stretchr/testify/require"
 )
 
 // WalkWithRunningServer walks path for datadriven files and calls RunTest on them.
@@ -45,6 +46,11 @@ func WalkWithNewServer(
 // RunTest executes PGTest commands, connecting to the database specified by
 // addr and user. Supported commands:
 //
+// "let": Run a query that returns a single row with a single column, and
+// save it in the variable named in the command argument. This variable can
+// be used in future "send" commands and is replaced by simple string
+// substitution.
+//
 // "send": Sends messages to a server. Takes a newline-delimited list of
 // pgproto3.FrontendMessage types. Can fill in values by adding a space then
 // a JSON object. No output. Messages with a []byte type (like CopyData) should
@@ -69,9 +75,11 @@ func WalkWithNewServer(
 // happens.
 func RunTest(t *testing.T, path, addr, user string) {
 	p, err := NewPGTest(context.Background(), addr, user)
+
 	if err != nil {
 		t.Fatal(err)
 	}
+	vars := make(map[string]string)
 	datadriven.RunTest(t, path, func(t *testing.T, d *datadriven.TestData) string {
 		switch d.Cmd {
 		case "only":
@@ -83,12 +91,40 @@ func RunTest(t *testing.T, path, addr, user string) {
 			}
 			return d.Expected
 
+		case "let":
+			require.Len(t, d.CmdArgs, 1, "only one argument permitted for let")
+			require.Truef(t, strings.HasPrefix(d.CmdArgs[0].Key, "$"), "let argument must begin with '$'")
+			lines := strings.Split(d.Input, "\n")
+			require.Len(t, lines, 1, "only one input command permitted for let")
+			require.Truef(t, strings.HasPrefix(lines[0], "Query "), "let must use a Query command")
+			if err := p.SendOneLine(lines[0]); err != nil {
+				t.Fatalf("%s: send %s: %v", d.Pos, lines[0], err)
+			}
+			msgs, err := p.Receive(hasKeepErrMsg(d), &pgproto3.DataRow{}, &pgproto3.ReadyForQuery{})
+			if err != nil {
+				t.Fatalf("%s: %+v", d.Pos, err)
+			}
+			sawData := false
+			for _, msg := range msgs {
+				if dataRow, ok := msg.(*pgproto3.DataRow); ok {
+					require.False(t, sawData, "let Query must return only one row")
+					require.Len(t, dataRow.Values, 1, "let Query must return only one column")
+					sawData = true
+					for _, arg := range d.CmdArgs {
+						vars[arg.Key] = string(dataRow.Values[0])
+					}
+				}
+			}
+			return ""
 		case "send":
 			if (d.HasArg("crdb_only") && !p.isCockroachDB) ||
 				(d.HasArg("noncrdb_only") && p.isCockroachDB) {
 				return d.Expected
 			}
 			for _, line := range strings.Split(d.Input, "\n") {
+				for k, v := range vars {
+					line = strings.ReplaceAll(line, k, v)
+				}
 				if err := p.SendOneLine(line); err != nil {
 					t.Fatalf("%s: send %s: %v", d.Pos, line, err)
 				}

